File: android/app/src/main/java/com/kairumo/padnote/canvas/SmartMagneticSnap.kt
package com.kairumo.padnote.canvas

import androidx.compose.ui.geometry.Offset
import kotlin.math.*

/**
 * 磁吸對齊結果
 */
data class MagneticSnapResult(
    val snappedPoint: Offset,
    val snappedAngleDegrees: Double?,
    val didSnap: Boolean
)

/**
 * 次世代筆跡磁吸對齊與幾何角度引導 (Smart Magnetic Snap & Ruler - Android 版)。
 * 當筆畫接近水平、垂直、45° 或紙張格線時，自動磁吸吸附並給予觸覺反饋與光導引。
 */
object SmartMagneticSnap {
    private val canonicalAngles = listOf(0.0, 45.0, 90.0, 135.0, 180.0, 225.0, 270.0, 315.0)
    private const val angleToleranceDegrees = 6.0
    private const val gridStep = 20f

    fun snap(start: Offset, current: Offset, enableGrid: Boolean = true): MagneticSnapResult {
        val dx = current.x - start.x
        val dy = current.y - start.y
        val distance = hypot(dx, dy)
        if (distance < 18f) {
            return MagneticSnapResult(current, null, false)
        }

        var angleDeg = Math.toDegrees(atan2(dy.toDouble(), dx.toDouble()))
        if (angleDeg < 0) angleDeg += 360.0

        var bestAngle: Double? = null
        for (canonical in canonicalAngles) {
            val diff = abs(angleDeg - canonical)
            val wrapDiff = min(diff, 360.0 - diff)
            if (wrapDiff <= angleToleranceDegrees) {
                bestAngle = canonical
                break
            }
        }

        if (bestAngle != null) {
            val rad = Math.toRadians(bestAngle)
            var snappedX = (start.x + cos(rad) * distance).toFloat()
            var snappedY = (start.y + sin(rad) * distance).toFloat()
            if (enableGrid) {
                snappedX = round(snappedX / gridStep) * gridStep
                snappedY = round(snappedY / gridStep) * gridStep
            }
            return MagneticSnapResult(
                Offset(snappedX, snappedY),
                bestAngle,
                true
            )
        }

        if (enableGrid) {
            val snappedX = round(current.x / gridStep) * gridStep
            val snappedY = round(current.y / gridStep) * gridStep
            if (abs(snappedX - current.x) < 5f && abs(snappedY - current.y) < 5f) {
                return MagneticSnapResult(Offset(snappedX, snappedY), null, true)
            }
        }

        return MagneticSnapResult(current, null, false)
    }
}

File: apple/Sources/SmartMagneticSnap.swift
//
//  SmartMagneticSnap.swift
//  Kairumo
//
//  次世代筆跡磁吸對齊與幾何角度引導 (Smart Magnetic Snap & Ruler)。
//  當筆畫接近水平、垂直、45° 或紙張格線時，自動磁吸吸附並給予觸覺反饋與光導引。
//

import CoreGraphics
import SwiftUI

public struct MagneticSnapResult: Equatable {
    public let snappedPoint: CGPoint
    public let snappedAngleDegrees: Double?
    public let didSnap: Bool

    public init(snappedPoint: CGPoint, snappedAngleDegrees: Double?, didSnap: Bool) {
        self.snappedPoint = snappedPoint
        self.snappedAngleDegrees = snappedAngleDegrees
        self.didSnap = didSnap
    }
}

public enum SmartMagneticSnap {
    private static let canonicalAngles: [Double] = [0, 45, 90, 135, 180, 225, 270, 315]
    private static let angleToleranceDegrees: Double = 6.0
    private static let gridStep: CGFloat = 20.0

    public static func snap(start: CGPoint, current: CGPoint, enableGrid: Bool = true) -> MagneticSnapResult {
        let dx = current.x - start.x
        let dy = current.y - start.y
        let distance = hypot(dx, dy)
        guard distance > 18 else {
            return MagneticSnapResult(snappedPoint: current, snappedAngleDegrees: nil, didSnap: false)
        }

        let angleRad = atan2(dy, dx)
        var angleDeg = angleRad * 180.0 / .pi
        if angleDeg < 0 { angleDeg += 360 }

        var bestAngle: Double? = nil
        for canonical in canonicalAngles {
            let diff = abs(angleDeg - canonical)
            let wrapDiff = min(diff, 360 - diff)
            if wrapDiff <= angleToleranceDegrees {
                bestAngle = canonical
                break
            }
        }

        if let angle = bestAngle {
            let rad = angle * .pi / 180.0
            var snappedX = start.x + cos(rad) * distance
            var snappedY = start.y + sin(rad) * distance
            if enableGrid {
                snappedX = round(snappedX / gridStep) * gridStep
                snappedY = round(snappedY / gridStep) * gridStep
            }
            return MagneticSnapResult(
                snappedPoint: CGPoint(x: snappedX, y: snappedY),
                snappedAngleDegrees: angle,
                didSnap: true
            )
        }

        if enableGrid {
            let snappedX = round(current.x / gridStep) * gridStep
            let snappedY = round(current.y / gridStep) * gridStep
            if abs(snappedX - current.x) < 5 && abs(snappedY - current.y) < 5 {
                return MagneticSnapResult(
                    snappedPoint: CGPoint(x: snappedX, y: snappedY),
                    snappedAngleDegrees: nil,
                    didSnap: true
                )
            }
        }

        return MagneticSnapResult(snappedPoint: current, snappedAngleDegrees: nil, didSnap: false)
    }
}
